### PR TITLE
[TorchFX] Rensen18 example

### PR DESCRIPTION
**Pull request by daniil-lyakhov**
### Changes

* Resnet18 TorchFX example

### Reason for changes

* To showcase NNCF TorchFX quantization

### Related tickets

#2766 

### Tests

test_examples/544/ - Done

### CODE REVIEW

File: examples/post_training_quantization/torch/fx/resnet18/main.py [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+from fastdownload import FastDownload
+from torch.jit import TracerWarning
+
+import nncf
+import nncf.torch
+from nncf.common.logging.track_progress import track
+from nncf.common.utils.helpers import create_table
+from nncf.torch.dynamic_graph.patch_pytorch import unpatch_torch_operators
+
+unpatch_torch_operators()
```
Comment (alexsu52) [inline examples/post_training_quantization/torch/fx/resnet18/main.py:36]: I would suggest using `with disable_patching():` instead of `unpatch_torch_operators()`.

File: examples/post_training_quantization/torch/fx/resnet18/main.py [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+
+warnings.filterwarnings("ignore", category=TracerWarning)
+warnings.filterwarnings("ignore", category=UserWarning)
```
Comment (alexsu52) [inline examples/post_training_quantization/torch/fx/resnet18/main.py:39]: Could you briefly explain why you ignore some warnings? Are these nncf warnings?

File: examples/post_training_quantization/torch/fx/resnet18/main.py [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+            to_image_dir.mkdir()
+        to_image_filepath = to_image_dir / image_filename
+        from_image_filepath.rename(to_image_filepath)
+    val_annotations_file.unlink()
+    val_images_dir.rmdir()
+
+
+def main():
+    device = torch.device("cpu")
+    print(f"Using {device} device")
```
Comment (alexsu52) [inline examples/post_training_quantization/torch/fx/resnet18/main.py:175]: Can I use GPU for quantization?

File: examples/post_training_quantization/torch/fx/resnet18/main.py [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+    quantized_fx_model = nncf.quantize(fx_model, quantization_dataset)
+
+    quantized_fx_model = torch.compile(quantized_fx_model, backend="openvino")
+
+    acc1_int8 = validate(val_loader, quantized_fx_model, device)
+    print(f"Accuracy@1 of INT8 model: {acc1_int8:.3f}")
+
+    ###############################################################################
+    # Step 3: Run benchmarks
+    print(os.linesep + "[Step 3] Run benchmarks")
```
Comment (alexsu52) [inline examples/post_training_quantization/torch/fx/resnet18/main.py:210]: I would suggest you explicitly write that you are testing torch.compile with the OpenVINO backend.

File: examples/post_training_quantization/torch/fx/resnet18/main.py [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+    fx_model = torch.export.export(model.eval(), args=(example_input,)).module()
+    quantized_fx_model = nncf.quantize(fx_model, quantization_dataset)
+
+    quantized_fx_model = torch.compile(quantized_fx_model, backend="openvino")
+
+    acc1_int8 = validate(val_loader, quantized_fx_model, device)
+    print(f"Accuracy@1 of INT8 model: {acc1_int8:.3f}")
+
+    ###############################################################################
+    # Step 3: Run benchmarks
```
Comment (alexsu52) [inline examples/post_training_quantization/torch/fx/resnet18/main.py:209]: What do you think about adding performance measurement of torch.compile with default backend?

File: examples/post_training_quantization/torch/fx/resnet18/main.py (hunk 3 above)
Comment (daniil-lyakhov) [reply examples/post_training_quantization/torch/fx/resnet18/main.py:167]: Unfortunately, I have found a bug for resnet18 model (ticket 154561). Besides, looks like GPU Graph module is not fully supported by torch compile in openvino (GPU batchnorm is not supported yet)

File: examples/post_training_quantization/torch/fx/resnet18/main.py (hunk 2 above)
Comment (daniil-lyakhov) [reply examples/post_training_quantization/torch/fx/resnet18/main.py]: Warnings are removed, thanks! There are no warning in this torchFX sample

File: examples/post_training_quantization/torch/fx/resnet18/main.py (hunk 1 above)
Comment (daniil-lyakhov) [reply examples/post_training_quantization/torch/fx/resnet18/main.py]: Done, please check

File: examples/post_training_quantization/torch/fx/resnet18/main.py (hunk 5 above)
Comment (daniil-lyakhov) [reply examples/post_training_quantization/torch/fx/resnet18/main.py:201]: Done

File: examples/post_training_quantization/torch/fx/resnet18/main.py (hunk 4 above)
Comment (daniil-lyakhov) [reply examples/post_training_quantization/torch/fx/resnet18/main.py:202]: Done, please check

File: examples/post_training_quantization/torch/fx/resnet18/main.py (hunk 5 above)
Comment (alexsu52) [reply examples/post_training_quantization/torch/fx/resnet18/main.py:201]: I mean adding measurement of FP32 model with torch.compile with default backend but not replacing torch.compile with OpenVINO backend.

File: examples/post_training_quantization/torch/fx/resnet18/main.py [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+    print(os.linesep + "[Step 3] Run benchmarks")
+    print("Run benchmark for FP32 model compiled with default backend ...")
+    compiled_model = torch.compile(model)
+    fp32_latency = measure_latency(compiled_model, example_inputs=example_input)
+
+    print("Run benchmark for INT8 model compiled with OpenVino backend ...")
```
Comment (alexsu52) [inline examples/post_training_quantization/torch/fx/resnet18/main.py:207]:
```suggestion
    print("Run benchmark for INT8 model compiled with openvino backend ...")
```

File: examples/post_training_quantization/torch/fx/resnet18/requirements.txt [hunk 7]
```diff
@@ -0,0 +1,4 @@
+fastdownload==0.0.7
+openvino==2024.3
```
Comment (alexsu52) [inline examples/post_training_quantization/torch/fx/resnet18/requirements.txt:2]:
```suggestion
openvino==2024.4
```

File: examples/post_training_quantization/torch/fx/resnet18/main.py (hunk 5 above)
Comment (daniil-lyakhov) [reply examples/post_training_quantization/torch/fx/resnet18/main.py:201]: I've updated the result table, please check
image

Comment (daniil-lyakhov) [reply examples/post_training_quantization/torch/fx/resnet18/main.py:201]: Update: 
image

File: examples/post_training_quantization/torch/fx/resnet18/main.py [hunk 8]
```diff
@@ -0,0 +1,225 @@
...
+    compiled_model = torch.compile(model, backend="openvino")
+    fp32_ov_latency = measure_latency(compiled_model, example_inputs=example_input)
+    print(f"{fp32_ov_latency:.3f} ms")
+    print(f"Speed up relative to FP32 model with default backend: x{fp32_latency / fp32_ov_latency:.3f}")
+
+    print("Run benchmark for INT8 model compiled with openvino backend ...")
+    int8_latency = measure_latency(quantized_fx_model, example_inputs=example_input)
+    print(f"{int8_latency:.3f} ms")
+    print(f"Speed up relative to FP32 model with default backend: x{fp32_latency / int8_latency:.3f}")
+    print(f"Speed up relative to FP32 model with openvino backend: x{fp32_ov_latency / int8_latency:.3f}")
```
Comment (alexsu52) [inline examples/post_training_quantization/torch/fx/resnet18/main.py:218]:
```suggestion
    print("Benchmark FP32 model compiled with default backend ...")
    compiled_model = torch.compile(model)
    fp32_latency = measure_latency(compiled_model, example_inputs=example_input)
    print(f"Latency:{fp32_latency:.3f} ms")

    print("Run benchmark for FP32 model compiled with openvino backend ...")
    compiled_model = torch.compile(model, backend="openvino")
    fp32_ov_latency = measure_latency(compiled_model, example_inputs=example_input)
    print(f"Latency: {fp32_ov_latency:.3f} ms")

    print("Run benchmark for INT8 model compiled with openvino backend ...")
    int8_latency = measure_latency(quantized_fx_model, example_inputs=example_input)
    print(f"Latency: {int8_latency:.3f} ms")
    
    print("[Step 4] torch.compile(..., backend="openvino") speed ups:"
    print(f"FP32 model: x{fp32_latency / fp32_ov_latency:.3f}")
    print(f"INT8 model: x{fp32_ov_latency / int8_latency:.3f}")
```

Comment (daniil-lyakhov) [reply examples/post_training_quantization/torch/fx/resnet18/main.py]: I fixed minor mistakes and aligned all prints as it done with `Benchmark FP32 model compiled with default backend ...`
Please check

**Review by AlexanderDokuchaev — CHANGES_REQUESTED** [review]

File: examples/post_training_quantization/torch/fx/resnet18/main.py (hunk 1 above)
Comment (alexsu52) [reply examples/post_training_quantization/torch/fx/resnet18/main.py]: Please move `with disable_patching():` to `def main():` for explicitly demonstration of using this one.

File: examples/post_training_quantization/torch/fx/resnet18/README.md [hunk 9]
```diff
@@ -0,0 +1,30 @@
+# Post-Training Quantization of Resnet18 PyTorch Model exported to torch.fx.GraphModule
```
[the remark this answers is not in the document]
Comment (alexsu52) [reply examples/post_training_quantization/torch/fx/resnet18/README.md:1]: I agree with `torch_fx/resenet18` in examples.

File: examples/post_training_quantization/torch/fx/resnet18/main.py [hunk 10]
```diff
@@ -0,0 +1,225 @@
...
+    print("Benchmark FP32 model compiled with openvino backend ...")
+    compiled_model = torch.compile(model, backend="openvino")
+    fp32_ov_latency = measure_latency(compiled_model, example_inputs=example_input)
+    print(f"{fp32_ov_latency:.3f} ms")
+
+    print("Benchmark INT8 model compiled with openvino backend ...")
+    int8_latency = measure_latency(quantized_fx_model, example_inputs=example_input)
+    print(f"{int8_latency:.3f} ms")
+
+    print('[Step 4] torch.compile(..., backend="openvino") speed ups:')
```
[the remark this answers is not in the document]
Comment (alexsu52) [reply examples/post_training_quantization/torch/fx/resnet18/main.py:217]: | Backend | Precision | Performance (ms) | Speed up | 
| - | - | -| - |
| default | FP32 | 3.2  |  | 
| openvino | FP32 | 1.4 |  x2 |
| openvino | INT8 | 0.6 | x4  |

File: examples/post_training_quantization/torch/fx/resnet18/main.py [hunk 11]
```diff
@@ -0,0 +1,225 @@
...
+        if not to_image_dir.exists():
+            to_image_dir.mkdir()
+        to_image_filepath = to_image_dir / image_filename
+        from_image_filepath.rename(to_image_filepath)
+    val_annotations_file.unlink()
+    val_images_dir.rmdir()
+
+
+def main():
+    device = torch.device("cpu")
```
[the remark this answers is not in the document]
Comment (daniil-lyakhov) [reply examples/post_training_quantization/torch/fx/resnet18/main.py]: It can only use GPU during statistic collection step, validation and performance measurements should be on CPU

File: examples/post_training_quantization/torch/fx/resnet18/README.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (daniil-lyakhov) [reply examples/post_training_quantization/torch/fx/resnet18/README.md]: Done

File: examples/post_training_quantization/torch/fx/resnet18/main.py [hunk 12]
```diff
@@ -0,0 +1,225 @@
...
+def get_resnet18_model(device: torch.device) -> torch.nn.Module:
+    num_classes = 200  # 200 is for Tiny ImageNet, default is 1000 for ImageNet
+    model = models.resnet18(weights=None)
+    # Update the last FC layer for Tiny ImageNet number of classes.
+    model.fc = nn.Linear(in_features=512, out_features=num_classes, bias=True)
+    model.to(device)
+    return model
+
+
+def measure_latency(model, example_inputs, num_iters=2000):
```
[the remark this answers is not in the document]
Comment (daniil-lyakhov) [reply examples/post_training_quantization/torch/fx/resnet18/main.py]: Done, thanks

File: examples/post_training_quantization/torch/fx/resnet18/main.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (daniil-lyakhov) [reply examples/post_training_quantization/torch/fx/resnet18/main.py]: image

File: examples/post_training_quantization/torch/fx/resnet18/main.py (hunk 1 above)
Comment (daniil-lyakhov) [reply examples/post_training_quantization/torch/fx/resnet18/main.py]: Done, please check

**Review by alexsu52 — APPROVED** [review]

File: examples/post_training_quantization/torch/fx/resnet18/main.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (AlexanderDokuchaev) [reply examples/post_training_quantization/torch/fx/resnet18/main.py]: `torch.compile(..., backend="openvino") speed ups`  still sounds like speed up of compilation.

It's should be `summary` (but need add accuracy here) or part of benchmark step.

**Comment by AlexanderDokuchaev** [comment]
@daniil-lyakhov Please rebase to develop and try run https://github.com/openvinotoolkit/nncf/actions/workflows/examples.yml workflow on your PR. 
And after need add duration of test to https://github.com/openvinotoolkit/nncf/blob/develop/tests/cross_fw/examples/.test_durations

File: examples/post_training_quantization/torch/fx/resnet18/main.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (daniil-lyakhov) [reply examples/post_training_quantization/torch/fx/resnet18/main.py]: @AlexanderDokuchaev, please specify exactly what is required from the table and the table description, ones for all

**Comment by daniil-lyakhov** [comment]
> @daniil-lyakhov Please rebase to develop and try run https://github.com/openvinotoolkit/nncf/actions/workflows/examples.yml workflow on your PR. And after need add duration of test to https://github.com/openvinotoolkit/nncf/blob/develop/tests/cross_fw/examples/.test_durations

Done:
https://github.com/openvinotoolkit/nncf/actions/runs/11381794559

File: examples/post_training_quantization/torch/fx/resnet18/main.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (daniil-lyakhov) [reply examples/post_training_quantization/torch/fx/resnet18/main.py]: @AlexanderDokuchaev please take a look

**Review by AlexanderDokuchaev — APPROVED** [review]
